### PR TITLE
Add ability to change data folder path

### PR DESCRIPTION
**Pull request by peppy**
- [x] Depends on https://github.com/ppy/osu/pull/9027

### CODE REVIEW

**Comment by MATRIX-feather** [comment]
In osu.Game/IO/OsuStorage.cs: line 19 :
``````
internal static readonly string[] IGNORE_DIRECTORIES = { "cache" }
``````
I think `osu-lazer` should be added to `IGNORE_DIRECTORIES` to avoid osu from getting stuck into a loop and finally make the disk runs out of space if the user selects the current data folder as a new location.

**Comment by peppy** [comment]
there's already a check for current directory. could you please provide reproduction for this?

**Comment by MATRIX-feather** [comment]
> there's already a check for current directory. could you please provide reproduction for this?

@peppy 
[simplescreenrecorder-2020-05-14_21.03.56.zip](https://github.com/ppy/osu/files/4628532/simplescreenrecorder-2020-05-14_21.03.56.zip)

**Comment by peppy** [comment]
thanks! that clears things up

File: osu.Game/IO/OsuStorage.cs [hunk 1]
```diff
@@ -48,11 +48,17 @@ public void Migrate(string newLocation)
             var source = new DirectoryInfo(GetFullPath("."));
             var destination = new DirectoryInfo(newLocation);
 
+            if (source.FullName == destination.FullName)
+                throw new ArgumentException("Destination provided is already the current location", nameof(newLocation));
+
+            if (destination.FullName.Contains(source.FullName))
```
Comment (bdach) [inline osu.Game/IO/OsuStorage.cs:54]: This check is broken if there are two sibling directories
```
C:/somedir
C:/somedir2
```
as the test failure demonstrates (the paths returned by `FullName` don't have trailing slashes - at least on Windows, therefore there is a wrongly-matching common prefix).

File: osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                },
+            };
+
+            Beatmap.Value = Beatmap.Default;
+
+            migrationTask = Task.Run(PerformMigration)
+                                .ContinueWith(t =>
+                                {
+                                    if (t.IsFaulted)
+                                        Logger.Log($"Error during migration: {t.Exception?.Message}", level: LogLevel.Error);
```
Comment (bdach) [inline osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs:91]: Arguably should be

```suggestion
                                        Logger.Error(t.Exception, "Error during migration: {t.Exception?.Message}");
```

as the current version completely discards stack traces, which means we won't be able to debug user reports in any of this.

**Review by bdach — DISMISSED** [review]

**Comment by bdach** [comment]
I've tried this a couple of times on Windows and it doesn't work too robustly. For example, you can start the migration process with imports running, leading to this occurring:

image

but I also think I got it to do that once after the import was already cancelled (I'm pretty sure). It might mean there's some kind of connection leak during the import process (possibly related to the leftover `client.db` files I've seen in #8420?)

I also got this once:

```
2020-05-14 20:51:41 [error]: System.AggregateException: One or more errors occurred. (The process cannot access the file '\\?\D:\lazertest\logs' because it is being used by another process.)
2020-05-14 20:51:41 [error]: ---> System.IO.IOException: The process cannot access the file '\\?\D:\lazertest\logs' because it is being used by another process.
2020-05-14 20:51:41 [error]: at System.IO.FileSystem.RemoveDirectoryInternal(String fullPath, Boolean topLevel, Boolean allowDirectoryNotEmpty)
2020-05-14 20:51:41 [error]: at System.IO.FileSystem.RemoveDirectoryRecursive(String fullPath, WIN32_FIND_DATA& findData, Boolean topLevel)
2020-05-14 20:51:41 [error]: at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
2020-05-14 20:51:41 [error]: at System.IO.DirectoryInfo.Delete(Boolean recursive)
2020-05-14 20:51:41 [error]: at osu.Game.IO.OsuStorage.deleteRecursive(DirectoryInfo target, Boolean topLevelExcludes) in D:\Open Source\osu\osu.Game\IO\OsuStorage.cs:line 91
2020-05-14 20:51:41 [error]: at osu.Game.IO.OsuStorage.Migrate(String newLocation) in D:\Open Source\osu\osu.Game\IO\OsuStorage.cs:line 73
2020-05-14 20:51:41 [error]: at osu.Game.OsuGameBase.Migrate(String path) in D:\Open Source\osu\osu.Game\OsuGameBase.cs:line 364
```

Haven't debugged too much today as my time is running out for the day so I just wanted to see how well it works for now. Might have a deeper look at it on the weekend.

**Comment by peppy** [comment]
yep, it’s not perfect and likely never will be. can probably add automatic retry logic in the future.

another plan is to have it lock the ability for database connections to be used during the process (after the connection flush)

File: osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs:91]: Using `Logger.Error` actually fires the error to our server handler right now, so we avoid using it in user scenarios like this. Maybe not the best handling there.

File: osu.Game/IO/OsuStorage.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/IO/OsuStorage.cs:54]: Fixed in #9027

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**